### ISSUE TITLE
Close/Open Index Issues

### ISSUE THREAD

**Issue by ebradshaw** [elastic/elasticsearch#14722]
I'm having an issue where closing and opening an index in an integration test appears to fail to properly recover the indexed documents:

https://gist.github.com/ebradshaw/a4fa97b7eac4257c4a2e

Is this an issue, or am I missing something?

**Comment by ebradshaw** [comment]
Also, this is appearing in both 1.4.2 and 1.7.3.

**Comment by dadoonet** [comment]
I think you should wait for a yellow status before sending your search request.

BTW you did not tell which exception you are getting and Es version.

Does not look like an issue to me.

**Comment by ebradshaw** [comment]
Thanks for the quick reply, David.

I'm tried this with both ES version 1.4.2 and 1.7.3.  Regarding the status, as I understand it ensureSearchable(...) verifies that the index status is GREEN, and I call this before doing the final count.  

I've also added a _refresh call into the test case, as that was necessary for the first assertion to pass:

https://gist.github.com/ebradshaw/a4fa97b7eac4257c4a2e

Regardless, the last assertion fails with:

java.lang.AssertionError: expected:<1> but was:<0>
	at __randomizedtesting.SeedInfo.seed([F88CF07515CF7DEF:C81F9BF1E097BCED]:0)
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.elasticearth.plugin.aggregation.CloseOpenIT.testCloseOpen(CloseOpenIT.java:39)

**Comment by ebradshaw** [comment]
@dadoonet It looks like the old index is deleted and a new one is created when the index is reopened.  When I look at the actual files in the index folder for the shard, all of the data files disappear and the segments_2 file is reverted to segments_1.  Does this still seem normal?

**Comment by dadoonet** [comment]
Wondering if you should not call [flush](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-flush.html) before closing the index?

**Comment by ebradshaw** [comment]
@dadoonet Tried calling flush and updated the test case.  It didn't seem to make a difference.  

I also ran a similar test using a transport client connected to node running in a separate JVM.  It passed!  Maybe the issue is with the embedded node?

**Comment by ebradshaw** [comment]
@dadoonet Unless I'm missing something, do you think we could reopen this issue?  While I can manage a workaround with a separate JVM, this isn't exactly test friendly :)

**Comment by rjernst** [comment]
@ebradshaw Have you tried looking at existing tests in ES? This test does what you want I believe:
https://github.com/elastic/elasticsearch/blob/6e81b0dd7025fe10d49ade35b5852697ef39355e/core/src/test/java/org/elasticsearch/indices/state/OpenCloseIndexIT.java#L329

My only thought is it looks like you are running on an old version based on the test constructs. What version would that be? If it is before 2.0 (as it appears it to be), you could be getting hit with the translog being fsynced asynchronously.

**Comment by dadoonet** [comment]
I wrote this test and it works fine on 2.0:

```java
        client.prepareIndex("test", "type").setSource("foo", "bar").get();
        client.admin().cluster().prepareHealth("test").setWaitForYellowStatus().get();
        client.admin().indices().prepareClose("test").get();
        client.admin().indices().prepareOpen("test").get();
        client.admin().cluster().prepareHealth("test").setWaitForYellowStatus().get();
        long count = client.prepareCount("test").get().getCount();
        assertThat(count, is(1L));
```

**Comment by ebradshaw** [comment]
@dadoonet That works for me also.  Must be time to upgrade to 2.0!  

@rjernst I was wondering if it might be something with the translog.  Looks like that testOpenCloseWithDocs wasn't there prior to 2.0.

Thanks for the help here guys.  BTW, it took me a bit of trouble getting the test to work in 2.0.  The [docs] (https://www.elastic.co/guide/en/elasticsearch/reference/current/using-elasticsearch-test-classes.html) call out two dependencies , but I also found that I had to include hamcrest-all-1.3 + junit-4.12 with hamcrest-core excluded.  If I didn't do this, I was dealing with either NoClassDefFoundErrors or Jar Hell errors.
